versioned_hdf5/subchunk_map.py

```diff
@@ -0,0 +1,309 @@
+from __future__ import annotations
+
+import itertools
+from typing import TYPE_CHECKING, Iterable, Iterator
+
+import cython
+import numpy as np
+from cython import Py_ssize_t
+from ndindex import (
+    BooleanArray,
+    ChunkSize,
+    Integer,
+    IntegerArray,
+    Slice,
+    Tuple,
+    ndindex,
+)
+from numpy.typing import NDArray
+
+if TYPE_CHECKING:
+    # TODO import from typing and remove quotes (requires Python 3.10)
+    # TODO use type <name> = ... (requires Python 3.12)
+    from typing_extensions import TypeAlias
+
+    AnySlicer: TypeAlias = (
+        "slice | NDArray[np.intp] | NDArray[np.bool] | tuple[()] | Py_ssize_t"
+    )
+
+
+@cython.cfunc
+def _ceiling(a: Py_ssize_t, b: Py_ssize_t) -> Py_ssize_t:
+    """Returns ceil(a/b)"""
+    return -(-a // b)
+
+
+@cython.cfunc
+def _smallest(x: Py_ssize_t, a: Py_ssize_t, m: Py_ssize_t) -> Py_ssize_t:
+    """Find the smallest integer y >= x where y = a + k*m for whole k's
+    Assumes 0 <= a <= x and m >= 1.
+
+    a                  x    y
+    | <-- m --> | <-- m --> |
+    """
+    n: Py_ssize_t = _ceiling(x - a, m)
+    return a + n * m
+
+
+@cython.cfunc
+def _subindex_chunk_slice(
+    s_start: Py_ssize_t,
+    s_stop: Py_ssize_t,
+    s_step: Py_ssize_t,
+    c_start: Py_ssize_t,
+    c_stop: Py_ssize_t,
+) -> slice:
+    start: Py_ssize_t = max(c_start, s_start)
+    # Get the smallest lcm multiple of common that is >= start
+    start = _smallest(start, s_start % s_step, s_step)
+    # Finally, we need to shift start so that it is relative to index
+    start = (start - s_start) // s_step
+
+    stop: Py_ssize_t = min(c_stop, s_stop)
+    stop = _ceiling(stop - s_start, s_step) if stop > s_start else 0
+
+    return slice(int(start), int(stop), 1)
+
+
+@cython.cfunc
+def _subindex_slice_chunk(
+    s_start: Py_ssize_t,
+    s_stop: Py_ssize_t,
+    s_step: Py_ssize_t,
+    c_start: Py_ssize_t,
+    c_stop: Py_ssize_t,
+) -> slice:
```

Comment:
I know this is a private function, but it still might be useful to provide a comment like "Given a chunk convert a slice that is relative to the start of the chunk into an index of the original dataset." or similar for contributors who don't have the same level of context about what this is supposed to return.

Reply:
done